keps/sig-api-machinery/2839-in-use-protection/README.md

```diff
@@ -0,0 +1,766 @@
+Therefore, in the initial implementation, any users with delete permission will be able to delete resources with liens by specifying the `IgnoreLiens` API option.
+
+#### Subresource
+
+There is a use case to disallow a controller to write to its CRD except for the status subresource, but allow it to update finalizers/liens.
+To allow this use case, it is required to provide a feature to set a separate RBAC to update finalizers/liens.
+On the other hand, finalizers field is currently not subresourced.
+Therefore, it needs to provide this feature in a compatible way for finalizers, and it would be better to implement liens field in the same way to finalizers.
+
+To solve this issue, there is [another KEP](https://github.com/kubernetes/enhancements/issues/3285) to add a new knob for just patching only a specific field via an endpoint of a subresource, without modifying the field itself.
+In this way, a feature to set a separate RBAC can be implemented later without making liens field itself a subresource.
+
+Subresourcing `Liens` is a beta blocker and won't be included in the initial implementation.
+
+#### Unresolved issues
+
+- Decision: When you delete a namespace and an object in that namespace has a lien:
+  - Does the namespace and object get deleted anyway?
+    - [x] Yes.
+  - Does the namespace deletion get blocked (i.e., the delete request fails) if any object in the namespace has a lien?
+    - [x] No.
+  - Does the other objects in the namespace get deleted, but that not object, preventing the namespace deletion from completing?
+    - [x] No. If you delete the namespace, the content will be removed.
+  - Does the namespace deletion not proceed until all objects in the namespace are free of liens?
+    - [ ] No. If you delete the namespace, the content will be removed. Namespace lifecycle cleanup will not honor liens.
+    - [x] No. If you delete the namespace, the content will be removed. Namespace lifecycle cleanup will honor liens only to guarantee the deletion order inside a namespace (guarantee of the deletion order is a beta blocker).
+    - [ ] ~~Yes. Namespace lifecycle cleanup will honor liens. If you delete the namespace, it won't be deleted until all the content will be deleted.~~
+- Name:
+    - [x] "liens" is short and precise, but an unusual word
+    - [ ] Other candidates: InUse, deleteInhibitors, hold, lease, claim, deletionBlockers, protections, guards
+- Subresource:
+  - Do you need a special permission to add a lien?
+    - [ ] Yes. Liens can only be changed via a call to resource/liens subresource. Create/update permission for the resource/liens subresource is required.
+    - [ ] ~~No. Only create/update permission forthe resource is required.~~
+    - [x] Not always. Create/Update permission for resource or update permission for resource/liens subresource are required (subresourcing is a beta blocker).
```

Comment:
Flagging to myself.  This is mentioned throughout, but I think it's important for alpha.  If we don't do it in alpha, I lack confidence that we will do it before beta.

Reply:
I wont' ACK a beta without subresource, personally.  I think we don't do enough subresourcing.